src/matchbox/client/helpers/selector.py

```
@@ -1,78 +1,128 @@
+from matchbox.common.sources import Match, SourceConfig, SourceField
 
 
 class Selector(BaseModel):
-    """A selector to choose a source and optionally a subset of fields to select."""
+    """A selector to choose a source and optionally a subset of columns to select."""
 
     model_config = ConfigDict(arbitrary_types_allowed=True)
 
-    address: SourceAddress
-    fields: list[str] | None = None
-    engine: Engine
+    source: SourceConfig
+    fields: list[SourceField]
+
+    @field_validator("source", mode="after")
+    @classmethod
+    def ensure_credentials(cls: type[Self], source: SourceConfig) -> SourceConfig:
+        """Ensure that the source has credentials set."""
+        if not source.location.credentials:
+            raise ValueError("Source credentials are not set")
+        return source
+
+    @property
+    def qualified_identifier(self) -> str:
+        """Get the qualified identifier for the source."""
+        return self.source.name + "_" + self.source.identifier
+
+    @property
+    def qualified_fields(self) -> list[str]:
+        """Get the qualified fields for the source."""
+        return [self.source.name + "_" + field.name for field in self.fields]
+
+    @classmethod
+    def from_name_and_credentials(
+        cls: type[Self],
+        name: SourceResolutionName,
+        credentials: Any,
+        fields: list[str] | None = None,
+    ) -> "Selector":
+        """Create a Selector from a source name and location credentials.
+
+        Args:
+            name: The name of the source to select from
+            credentials: The credentials to use for the source
+            fields: A list of fields to select from the source
+        """
+        source = _handler.get_source_config(name=name)
+        selected_fields: list[SourceField] = []
+        if fields:
+            for index_field in source.index_fields:
+                if index_field.name in fields:
+                    selected_fields.append(index_field)
```

the handling of un-indexed fields seem to have regressed. If you create a selector from __init__, you can give it whatever fields you want. If you use this factory function, the fields you pass it will be silently ignored

---

Isn't the code in this method redundant given the selector validators?